lib/runnable.js

```diff
@@ -260,6 +271,15 @@ Runnable.prototype.run = function (fn) {
   var ctx = this.ctx;
   var finished;
   var emitted;
+  process.on('beforeExit', detectAsyncUnscheduled);
+  if (this._closeIntercept) {
+    process.removeListener('beforeExit', this._closeIntercept);
+  }
+  this._closeIntercept = detectAsyncUnscheduled;
+
+  function detectAsyncUnscheduled () {
+    done(new Error('After ' + (new Date() - start) + ' milliseconds there were no more asynchronous actions scheduled. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.'));
```

Comment:
are there any other cases than `--timeout 0` or `this.timeout(0)` where this error would happen?  if not, we can be more specific.